File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temp: List[int]) -> List[int]:
        stk=[]
        ans=[0]*(len(temp))
        
        i=len(temp)-1
        while i>=0:
            while stk and stk[-1][0] <= temp[i]:
                stk.pop()
            if stk:
                ans[i] = stk[-1][1]-i
            stk.append((temp[i], i))
            i-=1
        
        return ans
            
        

File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temp) {
        int n = temp.length;
        Stack<int[]> stk = new Stack<>();
        int[] ans = new int[n];
        for(int i=n-1; i>=0; --i){
            while(!stk.isEmpty() && stk.peek()[0] <= temp[i]) stk.pop();
            
            if(stk.isEmpty()) ans[i] = 0;
            else ans[i] = stk.peek()[1] - i;
            
            stk.push(new int[]{temp[i], i});
        }
        
        return ans;
    }
}